File: pinterest.rb
require 'httpclient'
require 'nokogiri'
require 'json'
require File.join(File.dirname(__FILE__), './http_request')

class PinterestBot < HTTPRequest
  # Constructor
  #
  # username - String  
  # password - String
  def initialize(email, password)
    super()
    @email = email
    @password = password
  end
  
  # Public: Get csrf token from cookies
  # 
  # Return a String value of csrf token
  def csrftoken
    _get_cookie_name "csrftoken"
  end
  
  # Public: Login to pinterest with authenticate information
  # This function should be called after initialize
  # 
  # Example
  #   bot = PinterestBot(email, password)
  #   bot.login
  # 
  # Raise exception if login failed       
  def login
    get "https://pinterest.com/login/"
    body = {
      :email => @email,
      :password => @password,
      :csrfmiddlewaretoken => csrftoken
    }
    res = post "https://pinterest.com/login/", body
    raise "Login failed" if res.content != ""
  end
 
  # Public: Get username of this user
  #
  # Return a String    
  def username 
    return @username if @username
    
    res = get "https://pinterest.com/me/"
    url = res.headers["Location"]
    url = url[0..-2] if url[-1] == '/'
    @username = url[(url.rindex('/') + 1)..-1]
  end

  # Public: Get list of board of this account
  # 
  # Return a Dictionary which 
  #   key - String board name
  #   value - String board id
  def boards
    return @boards if @boards
    
    @boards = {}
    res = get "http://pinterest.com/#{username}/"
    soup = Nokogiri::HTML(res.content)

    div_boards = soup.css("div[class=\"pin pinBoard\"]")
    div_boards.each do |board|
      name = board.css("a")[0].content
      board_id = board["id"][5..-1]
      @boards[name] = board_id
    end
    @boards
  end
  
  # Public: Create new board
  #
  # name - String name of new board
  #   
  # Return a Dictionary object which contains
  #   url - String ULR of new category
  #   status - String :success is expected
  #   name - String `category_name`
  #   id - String ID of new category
  def create_new_board(name)
    headers = {
      "X-CSRFToken" => csrftoken(),
      "X-Pinterest-Referrer" => "http://pinterest.com/",
      "X-Requested-With" => "XMLHttpRequest",
    }
    body = {
      :name => name,
      :pass_category => true,
    }  
    res = post "https://pinterest.com/board/create/", body, headers
    JSON.parse res.content
  end
 
  # Public: Delete an board with name 
  #
  # name - String name of boards
  def delete_board(name)
    res = delete "https://pinterest.com/#{username}/#{name}/settings/"
    JSON.parse res.content
  end

  # Public: Get id for board name
  # 
  # name - String name of board     
  # 
  # Return String id of board
  def get_board_id_for_name(name)
    if not boards.include?(name) then
      res = create_new_board(name)
      board_id = res[:id]
      boards[name] = board_id
    end

    boards[name]
  end

  # Public: Upload photo to pinterest
  #
  # data - Dictionary which contains
  #   :caption - String caption you want to put in image
  #   :path - String path of photo in machine
  #   :board - String name of board you want to put new photo in
  #
  # Return Dictionary which contains
  #   :status - String "success" expected
  #   :url - String with format "/pin/<pin_id>/"
  #   :message - String "posted." expected
  def create_pin(data)
    File.open(data[:path]) do |file|
      body = {
        "csrfmiddlewaretoken" => csrftoken,
        "board" => get_board_id_for_name(data[:board]),
        "details" => data[:caption],
        "link" => "",
        "img_url" => "",
        "tags" => "",
        "replies" => "",
        "buyable" => "",
        "img" => file
      }
      res = post "http://pinterest.com/pin/create/", body
      JSON.parse res.content
    end
  end
  
  # Public: Update pinterest avatar with image from photo_path
  #
  # photo_path - String path to avatar photo
  # 
  # Return Dictionary 
  def update_avatar(photo_path)
    # first go to https://pinterest.com/settings/ to get user information
    res = get "https://pinterest.com/settings/"
    soup = Nokogiri::HTML(res.content)
    File.open(photo_path) do |file| 
      body = {
        "csrfmiddlewaretoken" => csrftoken,
        "first_name" => soup.css("input#id_first_name")[0]["value"] || "",
        "last_name" => soup.css("input#id_last_name")[0]["value"] || "",
        "language" => soup.css("select#id_language option[selected=\"selected\"]")[0]["value"],
        "username" => soup.css("input#id_username")[0]["value"],
        "gender" => soup.css("input[name=\"gender\"][checked=\"checked\"]")[0]["value"],
        "about" => soup.css("textarea#id_about")[0].content,
        "location" => soup.css("input#id_location")[0]["value"] || "",
        "website" => soup.css("input#id_website")[0]["value"] || "",
        "img" => file,
      }
      res = post "https://pinterest.com/settings/"
      res.content
    end
  end

end

File: pinterest.py
#!/usr/bin/env python

import requests
from bs4 import BeautifulSoup
import logging

class PinterestBot(object):
    def __init__(self, username, password):
        self.username = username
        self.password = password
        self.cookies = {}
        self.boards = {}
        self.headers = {
            "Accept" : "*/*",
            "Accept-Charset" : "ISO-8859-1,utf-8;q=0.7,*;q=0.3",
            "Accept-Encoding" : "gzip,deflate,sdch",
            "Accept-Language" : "en-US,en;q=0.8",
            "Connection" : "keep-alive",
            "Content-Type" : "application/x-www-form-urlencoded; charset=UTF-8",
            "User-Agent" : "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_7_3) AppleWebKit/536.11 (KHTML, like Gecko) Chrome/20.0.1132.47 Safari/536.11",
            "X-Requested-With" : "XMLHttpRequest", 
        }
        self.logger = logging.getLogger("")
        self.logger.addHandler(logging.StreamHandler())
   
    def _get_request_info(self, method, url, data=None):
        """Return string represent for the request"""
        info  = "Method %s\n" % method
        info += "URL: %s\n" % url
        info += "Cookies: \n"
        for k, v in self.cookies.items():
            info += "    %s: %s\n" % (k, v)
        if data:
            info += "Payload: \n"
            for k, v in data.items():
                info += "    %s: %s\n" % (k, v)
        return info

    def _make_request(self, method, url, data=None, headers=None):
        """Make request to server. If request sucessful, apply cookie to new request
        
        @return :class: Request object   
        """
        if not data: data = {}
        res = None
        try:
            if method == "GET":
                res = requests.get(url, cookies=self.cookies)
            elif method == "POST":
                if headers == None:
                    res = requests.post(url, data, cookies=self.cookies) 
                else:
                    res = requests.post(url, data, cookies=self.cookies, headers=headers)
            elif method == "PUT":
                res = requests.put(url, data, cookies=self.cookies) 

        except Exception as e:
            self.logger.error("--------------------------------------")
            self.logger.error("Cannot make request to %s" % url)
            self.logger.error(self._get_request_info(method, url, data))
            raise e

        if not res.ok:
            self.logger.error("--------------------------------------")
            self.logger.error("Request cannot get good respons to %s" % url)
            self.logger.error(self._get_request_info(method, url, data))
            self.logger.error("Response: ")
            self.logger.error(res.content)
        
        for k, v in res.cookies.items():
            self.cookies[k] = v
        return res
    
    def get(self, url):
        """Wrapper for :_make_request function"""
        return self._make_request("GET", url)

    def post(self, url, data=None, headers=None):
        """Wrapper for :_make_request function"""
        if headers == None:
            return self._make_request("POST", url, data=data)
        else:
            return self._make_request("POST", url, data=data, headers=headers)

    def put(self, url, data=None):
        """Wrapper for :_make_request function"""
        return self._make_request("PUT", url, data=data)

    def login(self):
        self.get("https://pinterest.com/login/")
        token = self.cookies["csrftoken"]
        self.post("https://pinterest.com/login/", data = {
            "email" : self.username,
            "password" : self.password,
            "csrfmiddlewaretoken" : token        
        })
        self._get_list_boards()
    
    def _get_list_boards(self):
        soup = BeautifulSoup(self.get("https://pinterest.com/me").content)              
        div_boards = soup.find_all("div", {"class" : "pin pinBoard"})
        for board in div_boards:
            board_name = board.find("a").contents[0].encode("utf-8").strip()
            board_id = board.attrs["id"][5:]
            self.boards[board_name] = board_id

    def create_category(self, category_name):
        if category_name in self.boards:
            print "Category %s already created" % category_name
            return

        data = {
            "name" : category_name,
        }
        headers = {
            "X-CSRFToken" : self.cookies["csrftoken"],
            "X-Pinterest-Referrer" : "http://pinterest.com/",
            "X-Requested-With" : "XMLHttpRequest",
            "User-Agent" : "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_7_4) AppleWebKit/536.11 (KHTML, like Gecko) Chrome/20.0.1132.57 Safari/536.11"
        }
        res = self.post("https://pinterest.com/board/create/", data=data, headers=headers)
        board_id = res.json["id"]
        self.boards[category_name] = board_id
        return res.json
    
    def get_board_id(self, category_name):
        if category_name not in self.boards:
            self.create_category(category_name)
        return self.boards[category_name]

    def upload(self, category_name, description, photo_path):
        data = {
            "board" : self.get_board_id(category_name),
            "details" : description, 
            "link" : "",
            "tags" : "",
            "buyable" : "",
            "img_url" : "",
            "csrfmiddlewaretoken" : self.cookies["csrftoken"],
            "img" : open(photo_path)
        }
        res = self.post("http://pinterests.com/pin/create/", data=data)
        return res

if __name__ == "__main__":
    bot = PinterestBot("knt.code@gmail.com", "@123456@")
    bot.login()
    res = bot.upload("test new", "My description", "/DATA/Dropbox/MacImage/IMG_0167.JPG")
    print res.headers
